core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java

```diff
@@ -59,6 +61,35 @@ public void addBuffer(ReadableBuffer buffer) {
     compositeBuffer.close();
   }
 
+  @Override
+  public boolean shouldUseByteBuffer() {
+    for (ReadableBuffer buf : buffers) {
+      if (!buf.shouldUseByteBuffer()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public List<ByteBuffer> readByteBuffers(int length) {
+    checkReadable(length);
+    readableBytes -= length;
```

Comment:
Maybe better to decrement this within the loop so that the buffer state remains consistent in case of runtime exception?

Also how about a special case (should be common):
```java
if (buffers.size() == 1) {
  return (readableBytes == 0 ? buffers.poll() : buffers.peek()).readByteBuffers(length);
}
```